File: src/main/java/com/demonwav/mcdev/platform/bukkit/BukkitProjectConfiguration.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

/*
 * IntelliJ IDEA Bukkit Support Plugin
 *
 * Written by Kyle Wood (DemonWav)
 * http://demonwav.com
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.bukkit;

import com.demonwav.mcdev.buildsystem.BuildSystem;
import com.demonwav.mcdev.platform.ProjectConfiguration;
import com.demonwav.mcdev.platform.bukkit.data.LoadOrder;
import com.demonwav.mcdev.util.Util;

import com.intellij.ide.util.EditorHelper;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BukkitProjectConfiguration extends ProjectConfiguration {

    @Nullable public LoadOrder loadOrder = null;
    public final List<String> loadBefore = new ArrayList<>();
    public final List<String> dependencies = new ArrayList<>();
    public final List<String> softDependencies = new ArrayList<>();
    public String prefix = null;
    public String minecraftVersion;

    public boolean hasPrefix() {
        return this.prefix != null && !this.prefix.trim().isEmpty();
    }

    public boolean hasLoadBefore() {
        return listContainsAtLeastOne(this.loadBefore);
    }

    public void setLoadBefore(String string) {
        this.loadBefore.clear();
        Collections.addAll(this.loadBefore, commaSplit(string));
    }

    public boolean hasDependencies() {
        return listContainsAtLeastOne(dependencies);
    }

    public void setDependencies(String string) {
        this.dependencies.clear();
        Collections.addAll(this.dependencies, commaSplit(string));
    }

    public boolean hasSoftDependencies() {
        return listContainsAtLeastOne(this.softDependencies);
    }

    public void setSoftDependencies(String string) {
        this.softDependencies.clear();
        Collections.addAll(this.softDependencies, commaSplit(string));
    }

    public boolean hasWebsite() {
        return this.website != null && !this.website.trim().isEmpty();
    }

    @Override
    public void create(@NotNull Project project, @NotNull BuildSystem buildSystem, @NotNull ProgressIndicator indicator) {
        Util.runWriteTask(() -> {
            try {
                indicator.setText("Writing main class");
                // Create plugin main class
                VirtualFile file = buildSystem.getSourceDirectories().get(0);
                final String[] files = this.mainClass.split("\\.");
                final String className = files[files.length - 1];

                final String packageName = this.mainClass.substring(0, this.mainClass.length() - className.length() - 1);
                file = getMainClassDirectory(files, file);

                final VirtualFile mainClassFile = file.findOrCreateChildData(this, className + ".java");

                BukkitTemplate.applyMainClassTemplate(project, mainClassFile, packageName, className);
                final VirtualFile pluginYml = buildSystem.getResourceDirectories().get(0).findOrCreateChildData(this, "plugin.yml");
                BukkitTemplate.applyPluginDescriptionFileTemplate(project, pluginYml, this, buildSystem);

                // Set the editor focus on the main class
                final PsiFile mainClassPsi = PsiManager.getInstance(project).findFile(mainClassFile);
                if (mainClassPsi != null) {
                    EditorHelper.openInEditor(mainClassPsi);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/bukkit/BukkitProjectConfiguration.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

@file:Suppress("Duplicates")

package com.demonwav.mcdev.platform.bukkit

import com.demonwav.mcdev.buildsystem.BuildDependency
import com.demonwav.mcdev.buildsystem.BuildRepository
import com.demonwav.mcdev.buildsystem.BuildSystem
import com.demonwav.mcdev.platform.PlatformType
import com.demonwav.mcdev.platform.ProjectConfiguration
import com.demonwav.mcdev.platform.bukkit.data.LoadOrder
import com.demonwav.mcdev.util.runWriteTask
import com.intellij.ide.util.EditorHelper
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiManager

class BukkitProjectConfiguration(override var type: PlatformType) : ProjectConfiguration(), BukkitLikeConfiguration {

    data class BukkitData(
        val loadOrder: LoadOrder = LoadOrder.POSTWORLD,
        val loadBefore: MutableList<String> = mutableListOf(),
        val prefix: String = "",
        val minecraftVersion: String = ""
    )

    var data: BukkitData? = null

    override val dependencies = mutableListOf<String>()
    override val softDependencies = mutableListOf<String>()

    fun hasPrefix() = data?.prefix?.isNotBlank() == true

    fun hasLoadBefore() = listContainsAtLeastOne(data?.loadBefore)
    fun setLoadBefore(string: String) {
        data?.loadBefore?.let { loadBefore ->
            loadBefore.clear()
            loadBefore.addAll(commaSplit(string))
        }
    }

    override fun hasDependencies() = listContainsAtLeastOne(dependencies)
    override fun setDependencies(string: String) {
        dependencies.clear()
        dependencies.addAll(commaSplit(string))
    }

    override fun hasSoftDependencies() = listContainsAtLeastOne(softDependencies)
    override fun setSoftDependencies(string: String) {
        softDependencies.clear()
        softDependencies.addAll(commaSplit(string))
    }

    fun hasWebsite() = !base?.website.isNullOrBlank()

    override fun create(project: Project, buildSystem: BuildSystem, indicator: ProgressIndicator) {
        if (project.isDisposed) {
            return
        }

        val baseConfig = base ?: return
        val dirs = buildSystem.directories ?: return

        runWriteTask {
            indicator.text = "Writing main class"
            // Create plugin main class
            var file = dirs.sourceDirectory
            val files = baseConfig.mainClass.split(".").toTypedArray()
            val className = files.last()

            val packageName = baseConfig.mainClass.substring(0, baseConfig.mainClass.length - className.length - 1)
            file = getMainClassDirectory(files, file)

            val mainClassFile = file.findOrCreateChildData(this, "$className.java")

            BukkitTemplate.applyMainClassTemplate(project, mainClassFile, packageName, className)
            val pluginYml = dirs.resourceDirectory.findOrCreateChildData(this, "plugin.yml")
            BukkitTemplate.applyPluginDescriptionFileTemplate(project, pluginYml, this, buildSystem)

            // Set the editor focus on the main class
            PsiManager.getInstance(project).findFile(mainClassFile)?.let { mainClassPsi ->
                EditorHelper.openInEditor(mainClassPsi)
            }
        }
    }

    override fun setupDependencies(buildSystem: BuildSystem) {
        val mcVersion = data?.minecraftVersion ?: return
        when (type) {
            PlatformType.PAPER -> {
                buildSystem.repositories.add(BuildRepository(
                    "papermc-repo",
                    "https://papermc.io/repo/repository/maven-public/"
                ))
                buildSystem.dependencies.add(BuildDependency(
                    "com.destroystokyo.paper",
                    "paper-api",
                    "$mcVersion-R0.1-SNAPSHOT",
                    mavenScope = "provided",
                    gradleConfiguration = "compileOnly"
                ))
                addSonatype(buildSystem.repositories)
            }
            PlatformType.SPIGOT -> {
                spigotRepo(buildSystem.repositories)
                buildSystem.dependencies.add(BuildDependency(
                    "org.spigotmc",
                    "spigot-api",
                    "$mcVersion-R0.1-SNAPSHOT",
                    mavenScope = "provided",
                    gradleConfiguration = "compileOnly"
                ))
                addSonatype(buildSystem.repositories)
            }
            PlatformType.BUKKIT -> {
                spigotRepo(buildSystem.repositories)
                buildSystem.dependencies.add(BuildDependency(
                    "org.bukkit",
                    "bukkit",
                    "$mcVersion-R0.1-SNAPSHOT",
                    mavenScope = "provided",
                    gradleConfiguration = "compileOnly"
                ))
            }
            else -> {}
        }
    }

    private fun spigotRepo(list: MutableList<BuildRepository>) {
        list.add(BuildRepository(
            "spigotmc-repo",
            "https://hub.spigotmc.org/nexus/content/repositories/snapshots/"
        ))
    }
}
